Revise radial gradient section (#137)
googlefonts/colr-gradients-spec: colr-gradients-spec.md

   the color at ω, but only painting on the parts of the bitmap that have not
   yet been painted on by earlier circles in this step for this rendering of the
   gradient.

The algorithm provides results in various cases as follows:

* When both radii are 0 (r₀ = r₁ = 0), then r(ω) is always 0 and nothing is painted.
* If the centers of the circles are distinct, the radii of the circles are different, and neither circle is entirely contained within the radius of the other circle, then the resulting shape resembles a cone that is open to one side. The surface outside the cone is not painted. (See figure 5.13.)
* If the centers of the circles are distinct but the radii are the same, and neither circle is contained within the other, then the result will be a strip, similar to the flattened projection of a circular cylinder. The surface outside the strip is not painted. (See figure 5.14.)
* If the radii of the circles are different but one circle is entirely contained within the radius of the other circle, the gradient will radiate outward in all directions from the inner circle, and the entire surface will be painted. (See figure 5.16.)

> NOTE: In the case of one circle being entirely contained within the other, if the radius of the inner circle is zero, then the gradient will radiate from the center point. No extension of the color line beyond that circle will be painted. Thus, the loop of the algorithm can be shortened to exclude values of ω corresponding to the inner circle (either 0 or 1) or beyond.

Reviewer: As this applies more generally (not just for when circles are contained within each other) but for whenever the radius becomes zero and the algorithm already states "...for ω where r(ω) > 0" this seems clear to me without this note.

> > **_TBD: What should the expected behaviour be in the case of both circles exactly overlapping with r > 0? (Only the extensions get painted, but how?)_**

Reviewer: Minor and optional: instead of TBDs in the text, let's keep track of TODOs in issues rather than in text.

Author: Opened #150

![Radial gradients with same-sized circles appearing as a strip.](images/colr_radial_gradients_strip.png)

**Figure 5.14 Radial gradients with same-sized circles appearing as a strip.**

Because the rendering algorithm progresses ω in a particular direction, from
positive infinity to negative infinity, and because pixels are not re-painted as
ω progresses, the appearance will be affected by which circle is considered
circle 0 and which is circle 1. 

This is illustrated in figure 5.15, in which three radial gradients are shown.
The first is the same as the first gradient shown in figure 5.13, using the pad
extend mode. In this first gradient, circle 0 is the small circle, on the left. In
the second gradient of figure 5.15, the start and end circles are reversed: circle
0 is the large circle, on the right. The color line is kept the same, and so the
red end starts at circle 0, now on the right. In the third gradient, the order
of stops in the color line is also reversed to put red on the left.
The key difference to notice between the gradients in figure 5.15 is the way colors
are painted in the interior: when the two circles are not overlapping, the arcs

> of constant color are centered on circle 1.

Reviewer: Perhaps
```suggestion
of constant color are concentric to circle 1.
```
?

Author: No: I suggested concentric earlier, but as @Lorp rightly pointed out, the circles are not in general concentric. (The interpolated circles will only be concentric if c0 = c1.)

Author: 'Centered on circle 1" still doesn't seem quite right, though, so have revised wording.

arcs of constant color are complete circles.

![Cone-shaped radial gradients with circle 0 and circle 1 swapped: the arcs of constant color are centered on circle 1.](images/colr_radial_gradients_direction.png)

**Figure 5.15 For cone-shaped radial gradients, arcs of constant color are centered on circle 1.**

When one circle is contained within the other, the extension of the gradient beyond the larger circle will fill the entire surface. Colors in the areas inside the inner circle and outside the outer circle are determined by the extend mode. Figure 5.16 illustrates this for the different extend modes.

![Radial gradients with one circle contained within the other.](images/colr_radial_gradients_circle_within_circle.png)

**Figure 5.16 Radial gradients with one circle contained within the other.**

NOTE: A scale transformation (see 5.7.11.1.5) can flatten shapes to resemble lines. If a radial gradient is nested in the child sub-graph of a transformation that flattens the circles so that they are nearly lines, the centers may still be separated by some distance. In that case, a radial gradient would appear as a strip or a cone filled with a linear gradient.

> > **_TBD: We still need to specify required behaviour for the case in which the transform really flattens the two circles, and the centers, to a line._**

Reviewer: See above, suggest to track TBDs and TODOs in issues, rather than inline.

Author: Opened #150

**5.7.11.1.2.3 Radial gradients**

> A basic radial gradient provides gradation of colors that radiates in all directions from an origin. The radial gradients supported in COLR version 1 provide a superset of this functionality: color gradation is defined along a cylinder defined by two circles. In the general case, the circles can have different radii to create a cone. A basic radial gradient is formed by one circle having a radius of zero with center located inside the other circle.

Reviewer: See below, for brevity and avoiding building definitions such as "basic radial gradient" that are not strictly needed for this spec, I would suggest to move the third paragraph up here and either strip this section or move it to a note and shorten it.

> NOTE: Some graphics libraries or other specifications support “radial” gradients with the same capabilities described here, whereas in other contexts, a “radial” gradient might provide more limited capabilities. In some contexts, the type of gradient defined here is referred to as a “two point conical” gradient.

Author: See earlier reply: I think _something_ needs to be kept.

> A radial gradient is defined by two circles, one with center c₀ and radius r₀, and another with center c₁ and radius r₁, plus a color line. The color line is related to the two circles by aligning stop offset 0 with a point on the perimeriter of the first circle (with center c₀) and aligning stop offset 1.0 with a corresponding point (the same angular direction from the center) on the perimeter of the second circle.

Reviewer: I assume you did want to address the fact that depending on repeat mode, not the full circle is drawn with the color at the respective stop offset, is that the case?

It's good that you fixed and removed the remaining notion of `optionally transformed`.

I don't think the "alignment of stop offsets to a point on the perimeter of..." formulation adds precision here as the algorithm describes the pixel and color placement accurately. We only and mainly need to specify the direction or association of the color line to circles, I suggest instead:

```suggestion
A radial gradient is defined by two circles, one with center c₀ and radius r₀, and another with center c₁ and radius r₁, plus a color line. The color line is mapped to the two circles by associating stop offset 0 with first circle (with center c₀) and associating stop offset 1.0 with the second circle (with center c₁), with rules for which pixels to draw as in the algorithm below.
```

> A radial gradient is defined by two circles, one with center c₀ and radius r₀, and another with center c₁ and radius r₁, plus a color line. The color line is related to the two circles by aligning stop offset 0 with a point on the perimeriter of the first circle (with center c₀) and aligning stop offset 1.0 with a corresponding point (the same angular direction from the center) on the perimeter of the second circle.

Author: >I think it makes sense to use this as the first paragraph of this section instead of starting with a comparison and definition of what a "basic radial gradient" is intended to mean. We do not need to go into details what other spec, implementations or documentation calls their gradients. For brevity and conciseness, I think we should focus on what we define a radial gradient to mean in this spec.

Because the term "radial gradient" gets used in different contexts in different ways, I think there should be _some_ clarification to distinguish from other uses of the term. But it can be more concise.

> A radial gradient is defined by two circles, one with center c₀ and radius r₀, and another with center c₁ and radius r₁, plus a color line. The color line is related to the two circles by aligning stop offset 0 with a point on the perimeriter of the first circle (with center c₀) and aligning stop offset 1.0 with a corresponding point (the same angular direction from the center) on the perimeter of the second circle.

Author: >I assume you did want to address the fact that depending on repeat mode, not the full circle is drawn with the color at the respective stop offset, is that the case?

Not sure what you have in mind. Step 3 of the algorithm does say,

>but only painting on the parts of the bitmap that have not yet been painted on by earlier circles

Do you think this should be called out in prose?

> A radial gradient is defined by two circles, one with center c₀ and radius r₀, and another with center c₁ and radius r₁, plus a color line. The color line is related to the two circles by aligning stop offset 0 with a point on the perimeriter of the first circle (with center c₀) and aligning stop offset 1.0 with a corresponding point (the same angular direction from the center) on the perimeter of the second circle.

Author: Resolved comments except for the one mentioned in my last reply.

The drawing algorithm for radial gradients follows the [HTML WHATWG Canvas specification for
createRadialGradient()][32], but adapted with with alternate color line extend modes, as
described in 5.7.11.1.2.1. Radial gradients shall be rendered with results that match the
results produced by the following steps.

With circle center points c₀ and c₁ defined as c₀ = (x₀, y₀) and c₁ = (x₁, y₁):

1. If c₀ = c₁ and r₀ = r₁ then paint nothing and return.
2. For real values of ω:<br>
   Let x(ω) = (x₁-x₀)ω + x₀<br>
   Let y(ω) = (y₁-y₀)ω + y₀<br>
   Let r(ω) = (r₁-r₀)ω + r₀<br>
   Let the color at ω be the color at that position on the gradient color line.
3. For all values of ω where r(ω) > 0, starting with the value of ω nearest to
   positive infinity and ending with the value of ω nearest to negative
   infinity, draw the circumference of the ellipse resulting from translating

>    a circle with radius r(ω) by affine transform such that the center is at position (x(ω), y(ω)), with

Reviewer: ```suggestion
   a circle with radius r(ω) centered at position (x(ω), y(ω)), with
```

Suggest to strip notion of transform. I think that's also a leftover from when we had a transform parameter here.

   the color at ω, but only painting on the parts of the bitmap that have not
   yet been painted on by earlier circles in this step for this rendering of the
   gradient.

> Figure 5.13 illustrates a radial gradient using the three different color line extend modes. The color line is defined with stops for the interval [0, 1]. Note that the circles that define the gradient are not stroked as part of the gradient itself. Stroked circles have been overlaid in the figure to illustrate the color line and the region that is painted in relation to the two circles.

Reviewer: This paragraph, or the Figure caption now lacks a description of the colors on the color stops, that the previous text had: 

"color stops for blue at 0, yellow at 0.5 and red at 1."

circle is wholly contained within the other, however: in that case, the
of color always bends towards circle 0.

NOTE: This difference does not exist if one
circle is entirely contained within the other: in that case, the
individual colors are complete, concentric circles.

![Radial gradients with start and end circles swapped](images/radial_gradients_direction.png)
![Radial gradients with start and end circles swapped.](images/colr_radial_gradients_direction.png)

*__Figure 4:__ Radial gradients with start and end circles swapped. (Illustration
generated from <a href="images/radial_gradient_direction.svg">images/radial_gradient_direction.svg</a>*
**Figure 5.14 Radial gradients with start and end circles swapped.**

*__Note:__ It follows from the algorithm that: a) When both radii are 0, r₀ = r₁
= 0, then r(ω) is always 0 and nothing is painted. b) If the centers of the
circles are distinct, one has a zero radius and lies outside the radius of the
other, the other has a non-zero radius, the resulting shape resembles a cone
that is open to one side, see Figure 3 below. c) If the circles centers are
identical, one has a zero radius, one has a non-zero radius, then the gradient

> is well-defined and is painted according to the algorithm.*

Reviewer: It seems to me that these clarifications are dropped in the new text, but they were added to address the questions raised in an issue what happens in edge cases. I suggest to add those back in.

Author: The existing text didn't completely cover all cases (e.g., b is overly specific, and another important edge case is not mentioned). I'll re-incorporate revised text.
